File: packages/ssl-pinning/android/src/main/java/io/capkit/sslpinning/config/SSLPinningConfig.kt
package io.capkit.sslpinning.config

import com.getcapacitor.Plugin
import io.capkit.sslpinning.logger.SSLPinningLogger
import java.io.BufferedReader
import java.io.InputStreamReader

/**
 * Plugin configuration container.
 *
 * This class is responsible for reading and exposing
 * static configuration values defined under the
 * `SSLPinning` key in capacitor.config.ts.
 *
 * Configuration rules:
 * - Read once during plugin initialization
 * - Treated as immutable runtime input
 * - Accessible only from native code
 *
 * Validation:
 * - Certificate files are validated at load time
 * - Missing/invalid certs cause fail-fast errors
 */
class SSLPinningConfig(
  plugin: Plugin,
) {
  // -----------------------------------------------------------------------------
  // Properties
  // -----------------------------------------------------------------------------

  /**
   * Enables verbose native logging.
   *
   * When enabled, additional debug information
   * is printed to Logcat.
   *
   * Default: false
   */
  val verboseLogging: Boolean

  /**
   * Default SHA-256 fingerprint used by checkCertificate()
   * when no fingerprint is provided at runtime.
   */
  val fingerprint: String?

  /**
   * Default SHA-256 fingerprints used by checkCertificates()
   * when no fingerprints are provided at runtime.
   */
  val fingerprints: List<String>

  /**
   * Certificate filenames used for SSL pinning.
   *
   * Files are expected to be located under:
   * assets/certs/
   *
   * This is the global fallback used when no domain-specific
   * certificates are configured via `certsByDomain`.
   */
  val certs: List<String>

  /**
   * Per-domain certificate configuration.
   *
   * Maps a domain (or subdomain pattern) to a list of
   * certificate filenames to use for that domain.
   */
  val certsByDomain: Map<String, List<String>>

  /**
   * Optional manifest file for certificate auto-discovery.
   *
   * The manifest is a JSON file containing either:
   * - { "certs": ["a.cer", "b.cer"] }
   * - { "certsByDomain": { "example.com": ["cert.cer"] } }
   * - Both, which extend/override the explicit config values
   */
  val certsManifest: String?

  /**
   * Domains or URL prefixes excluded from SSL pinning.
   *
   * Any request whose host matches one of these values
   * MUST bypass SSL pinning checks.
   */
  val excludedDomains: List<String>

  /**
   * Cached validation status for certificate files.
   * Populated during initialization.
   */
  private val certValidationCache: Map<String, Boolean>

  // -----------------------------------------------------------------------------
  // Initialization
  // -----------------------------------------------------------------------------

  init {
    val config = plugin.getConfig()
    val context = plugin.context

    // Verbose logging flag
    verboseLogging =
      config.getBoolean("verboseLogging", false)

    SSLPinningLogger.verbose = verboseLogging

    // Single fingerprint (optional)
    val fp = config.getString("fingerprint")
    fingerprint =
      if (!fp.isNullOrBlank()) fp else null

    // Multiple fingerprints (optional)
    fingerprints =
      config
        .getArray("fingerprints")
        ?.toList()
        ?.mapNotNull { it as? String }
        ?.filter { it.isNotBlank() }
        ?: emptyList()

    // Certificate files (optional)
    val explicitCerts =
      config
        .getArray("certs")
        ?.toList()
        ?.mapNotNull { it as? String }
        ?.filter { it.isNotBlank() }
        ?: emptyList()

    // Per-domain certificates (optional)
    val explicitCertsByDomain =
      parseCertsByDomainFromConfig(config)

    // Manifest (optional)
    val manifestPath = config.getString("certsManifest")

    // Load and merge manifest if present
    val (manifestCerts, manifestCertsByDomain) =
      if (!manifestPath.isNullOrBlank()) {
        loadManifest(context, manifestPath)
      } else {
        Pair(emptyList(), emptyMap<String, List<String>>())
      }

    // Merge: explicit config takes precedence over manifest
    certs = explicitCerts.ifEmpty { manifestCerts }

    val mergedCertsByDomain = mutableMapOf<String, List<String>>()
    mergedCertsByDomain.putAll(manifestCertsByDomain)
    mergedCertsByDomain.putAll(explicitCertsByDomain)
    certsByDomain = mergedCertsByDomain.toMap()

    // Store manifest path
    certsManifest = manifestPath

    // Excluded domains (optional)
    excludedDomains =
      config
        .getArray("excludedDomains")
        ?.toList()
        ?.mapNotNull { it as? String }
        ?.filter { it.isNotBlank() }
        ?: emptyList()

    // Validate certificate files at load time (fail-fast)
    certValidationCache = validateCertFiles(context, certs, certsByDomain)
  }

  // -----------------------------------------------------------------------------
  // Helpers
  // -----------------------------------------------------------------------------

  /**
   * Checks if a certificate file exists and is valid.
   */
  fun isCertValid(fileName: String): Boolean = certValidationCache[fileName] ?: false

  /**
   * Returns all certificate file names from config (global + domain-specific).
   */
  fun getAllCertFileNames(): Set<String> {
    val domainCerts = certsByDomain.values.flatten()
    return (certs + domainCerts).toSet()
  }

  private fun parseCertsByDomainFromConfig(config: com.getcapacitor.PluginConfig): Map<String, List<String>> {
    val jsonObject = config.getObject("certsByDomain") as? org.json.JSONObject
    if (jsonObject == null) return emptyMap()

    val result = mutableMapOf<String, List<String>>()
    val keys = jsonObject.keys()

    while (keys.hasNext()) {
      val key = keys.next()
      val jsonArray = jsonObject.getJSONArray(key)
      val certList = mutableListOf<String>()
      for (i in 0 until jsonArray.length()) {
        val cert = jsonArray.getString(i)
        if (cert.isNotBlank()) {
          certList.add(cert)
        }
      }
      if (certList.isNotEmpty()) {
        result[key] = certList
      }
    }
    return result
  }

  private fun loadManifest(
    context: android.content.Context,
    manifestPath: String,
  ): Pair<List<String>, Map<String, List<String>>> {
    val assetPath = "certs/$manifestPath"

    return try {
      context.assets.open(assetPath).use { inputStream ->
        BufferedReader(InputStreamReader(inputStream)).use { reader ->
          val json = reader.readText()
          parseManifestJson(json)
        }
      }
    } catch (e: Exception) {
      SSLPinningLogger.error("Failed to load manifest: $assetPath", e)
      Pair(emptyList(), emptyMap())
    }
  }

  private fun parseManifestJson(json: String): Pair<List<String>, Map<String, List<String>>> =
    try {
      val result = org.json.JSONObject(json)
      val certsList = mutableListOf<String>()
      val certsByDomainMap = mutableMapOf<String, List<String>>()

      // Parse global certs
      if (result.has("certs")) {
        val certsArray = result.getJSONArray("certs")
        for (i in 0 until certsArray.length()) {
          val cert = certsArray.getString(i)
          if (cert.isNotBlank()) {
            certsList.add(cert)
          }
        }
      }

      // Parse certsByDomain
      if (result.has("certsByDomain")) {
        val domainObj = result.getJSONObject("certsByDomain")
        val keys = domainObj.keys()
        while (keys.hasNext()) {
          val domain = keys.next()
          val certArray = domainObj.getJSONArray(domain)
          val certList = mutableListOf<String>()
          for (i in 0 until certArray.length()) {
            val cert = certArray.getString(i)
            if (cert.isNotBlank()) {
              certList.add(cert)
            }
          }
          if (certList.isNotEmpty()) {
            certsByDomainMap[domain] = certList
          }
        }
      }

      Pair(certsList, certsByDomainMap)
    } catch (e: Exception) {
      SSLPinningLogger.error("Failed to parse manifest JSON", e)
      Pair(emptyList(), emptyMap())
    }

  /**
   * Validates all configured certificate files at load time.
   * Returns a map of filename -> isValid.
   */
  private fun validateCertFiles(
    context: android.content.Context,
    globalCerts: List<String>,
    domainCerts: Map<String, List<String>>,
  ): Map<String, Boolean> {
    val result = mutableMapOf<String, Boolean>()
    val allCerts = (globalCerts + domainCerts.values.flatten()).distinct()

    for (certFile in allCerts) {
      val isValid = isCertFileValid(context, certFile)
      result[certFile] = isValid
      if (!isValid) {
        SSLPinningLogger.error("Certificate validation failed: $certFile")
      }
    }

    return result
  }

  /**
   * Checks if a certificate file exists and can be parsed.
   */
  private fun isCertFileValid(
    context: android.content.Context,
    fileName: String,
  ): Boolean {
    val assetPath = "certs/$fileName"
    return try {
      context.assets.open(assetPath).use { inputStream ->
        val certFactory =
          java.security.cert.CertificateFactory
            .getInstance("X.509")
        val cert = certFactory.generateCertificate(inputStream)
        cert is java.security.cert.X509Certificate
      }
    } catch (e: Exception) {
      false
    }
  }
}

File: packages/ssl-pinning/ios/Sources/SSLPinningPlugin/config/SSLPinningConfig.swift
import Foundation
import Capacitor

/**
 Plugin configuration container.

 This struct is responsible for reading and exposing
 static configuration values defined under the
 `SSLPinning` key in capacitor.config.ts.

 Configuration rules:
 - Read once during plugin initialization
 - Treated as immutable runtime input
 - Accessible only from native code

 Validation:
 - Certificate files are validated at load time
 - Missing/invalid certs cause fail-fast errors
 */
public struct SSLPinningConfig {

    // MARK: - Configuration Keys

    /**
     Centralized definition of configuration keys.
     Avoids string duplication and typos.
     */
    private enum Keys {
        static let verboseLogging = "verboseLogging"
        static let fingerprint = "fingerprint"
        static let fingerprints = "fingerprints"
        static let certs = "certs"
        static let certsByDomain = "certsByDomain"
        static let certsManifest = "certsManifest"
        static let excludedDomains = "excludedDomains"
    }

    // MARK: - Public Configuration Values

    /**
     Enables verbose native logging.

     When enabled, the plugin prints additional
     debug information to the Xcode console.

     Default: false
     */
    public let verboseLogging: Bool

    /**
     Default SHA-256 fingerprint used by `checkCertificate()`
     when no fingerprint is provided at runtime.
     */
    public let fingerprint: String?

    /**
     Default SHA-256 fingerprints used by `checkCertificates()`
     when no fingerprints are provided at runtime.
     */
    public let fingerprints: [String]

    /**
     Certificate filenames used for SSL pinning.

     Files are expected to be copied into the app bundle
     subdirectory 'certs'.

     This is the global fallback used when no domain-specific
     certificates are configured via `certsByDomain`.
     */
    public let certs: [String]

    /**
     Per-domain certificate configuration.

     Maps a domain (or subdomain pattern) to a list of
     certificate filenames to use for that domain.
     */
    public let certsByDomain: [String: [String]]

    /**
     Optional manifest file for certificate auto-discovery.

     The manifest is a JSON file containing either:
     - { "certs": ["a.cer", "b.cer"] }
     - { "certsByDomain": { "example.com": ["cert.cer"] } }
     - Both, which extend/override the explicit config values
     */
    public let certsManifest: String?

    /**
     Domains or URL prefixes excluded from SSL pinning.

     Any request whose host matches one of these values
     MUST bypass SSL pinning checks.
     */
    public let excludedDomains: [String]

    // MARK: - Defaults

    private static let defaultVerboseLogging: Bool = false
    private static let defaultFingerprint: String? = nil
    private static let defaultFingerprints: [String] = []
    private static let defaultCerts: [String] = []
    private static let defaultCertsByDomain: [String: [String]] = [:]
    private static let defaultCertsManifest: String? = nil
    private static let defaultExcludedDomains: [String] = []

    // MARK: - Validation Cache

    private let certValidationCache: [String: Bool]

    // MARK: - Initialization

    /**
     Initializes the configuration by reading values
     from the Capacitor PluginConfig.

     - Parameter plugin: The CAPPlugin instance used
     to access typed configuration values.
     */
    init(plugin: CAPPlugin) {
        let config = plugin.getConfig()

        // Verbose logging flag
        self.verboseLogging =
            config.getBoolean(
                Keys.verboseLogging,
                Self.defaultVerboseLogging
            )

        // Synchronize logger state
        SSLPinningLogger.verbose = self.verboseLogging

        // Single fingerprint (optional)
        if let value = config.getString(Keys.fingerprint),
           !value.isEmpty {
            self.fingerprint = value
        } else {
            self.fingerprint = Self.defaultFingerprint
        }

        // Multiple fingerprints (optional)
        self.fingerprints =
            config.getArray(Keys.fingerprints)?
            .compactMap { $0 as? String }
            .filter { !$0.isEmpty }
            ?? Self.defaultFingerprints

        // Certificate files (optional)
        let explicitCerts: [String] =
            config.getArray(Keys.certs)?
            .compactMap { $0 as? String }
            .filter { !$0.isEmpty }
            ?? Self.defaultCerts

        // Per-domain certificates (optional)
        let explicitCertsByDomain = Self.parseCertsByDomain(config.getObject(Keys.certsByDomain))

        // Manifest (optional)
        let manifestPath = config.getString(Keys.certsManifest)

        // Load and merge manifest if present
        let (manifestCerts, manifestCertsByDomain) = Self.loadManifest(manifestPath)

        // Merge: explicit config takes precedence over manifest
        self.certs = explicitCerts.isEmpty ? manifestCerts : explicitCerts

        var mergedCertsByDomain = manifestCertsByDomain
        for (key, value) in explicitCertsByDomain {
            mergedCertsByDomain[key] = value
        }
        self.certsByDomain = mergedCertsByDomain

        // Store manifest path
        self.certsManifest = manifestPath

        // Excluded domains (optional)
        self.excludedDomains =
            config.getArray(Keys.excludedDomains)?
            .compactMap { $0 as? String }
            .filter { !$0.isEmpty }
            ?? Self.defaultExcludedDomains

        // Validate certificate files at load time (fail-fast)
        self.certValidationCache = Self.validateCertFiles(
            certs: self.certs,
            certsByDomain: self.certsByDomain
        )

        // Check for invalid certificates and log errors
        for (certFile, isValid) in self.certValidationCache {
            if !isValid {
                SSLPinningLogger.error("Certificate validation failed: \(certFile)")
            }
        }
    }

    // MARK: - Public Helpers

    /**
     Checks if a certificate file exists and is valid.
     */
    public func isCertValid(_ fileName: String) -> Bool {
        return certValidationCache[fileName] ?? false
    }

    /**
     Returns all certificate file names from config (global + domain-specific).
     */
    public func getAllCertFileNames() -> Set<String> {
        let domainCerts = certsByDomain.values.flatMap { $0 }
        return Set(certs + domainCerts)
    }

    // MARK: - Private Helpers

    private static func parseCertsByDomain(_ object: [String: Any]?) -> [String: [String]] {
        guard let obj = object else { return [:] }

        var result: [String: [String]] = [:]
        for (key, value) in obj {
            if let certArray = value as? [String] {
                let filtered = certArray.filter { !$0.isEmpty }
                if !filtered.isEmpty {
                    result[key] = filtered
                }
            }
        }
        return result
    }

    private static func loadManifest(_ manifestPath: String?) -> ([String], [String: [String]]) {
        guard let path = manifestPath, !path.isEmpty else {
            return ([], [:])
        }

        guard let url = Bundle.main.url(forResource: path.replacingOccurrences(of: ".json", with: ""), withExtension: "json"),
              let data = try? Data(contentsOf: url) else {
            SSLPinningLogger.error("Failed to load manifest: \(path)")
            return ([], [:])
        }

        return parseManifestJson(data)
    }

    private static func parseManifestJson(_ data: Data) -> ([String], [String: [String]]) {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            SSLPinningLogger.error("Failed to parse manifest JSON")
            return ([], [:])
        }

        var certsList: [String] = []
        var certsByDomainMap: [String: [String]] = [:]

        // Parse global certs
        if let certsArray = json["certs"] as? [String] {
            certsList = certsArray.filter { !$0.isEmpty }
        }

        // Parse certsByDomain
        if let domainObj = json["certsByDomain"] as? [String: [String]] {
            for (domain, certList) in domainObj {
                let filtered = certList.filter { !$0.isEmpty }
                if !filtered.isEmpty {
                    certsByDomainMap[domain] = filtered
                }
            }
        }

        return (certsList, certsByDomainMap)
    }

    /**
     Validates all configured certificate files at load time.
     Returns a map of filename -> isValid.
     */
    private static func validateCertFiles(
        certs: [String],
        certsByDomain: [String: [String]]
    ) -> [String: Bool] {
        var result: [String: Bool] = [:]
        let allCerts = Array(Set(certs + certsByDomain.values.flatMap { $0 }))

        for certFile in allCerts {
            let isValid = isCertFileValid(certFile)
            result[certFile] = isValid
        }

        return result
    }

    /**
     Checks if a certificate file exists and can be parsed.
     */
    private static func isCertFileValid(_ fileName: String) -> Bool {
        let fileURL = URL(fileURLWithPath: fileName)
        let name = fileURL.deletingPathExtension().lastPathComponent
        let ext = fileURL.pathExtension.isEmpty ? "cer" : fileURL.pathExtension

        guard let path = Bundle.main.path(forResource: name, ofType: ext),
              let data = try? Data(contentsOf: URL(fileURLWithPath: path)),
              let _ = SecCertificateCreateWithData(nil, data as CFData) else {
            return false
        }

        return true
    }
}
